### PR TITLE
[UNR-2987] Actor hierarchy migration

### PR DESCRIPTION
**Pull request by ImprobableNic**
#### Description
Actors belonging to the same ownership hierarchy depends on each other to work. This change makes sure that they all send a transition request together.
They could arrive at different time on their destination though, and could try to transition to an incorrect position. A transition backoff period has been added to mitigate this issue.
Edge cases remains related to snapshot or worker crash, and would need actor sets to be properly handled.

#### Release note
TODO

#### Tests
TODO : Teleporting Gym.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h [hunk 1]
```diff
@@ -359,4 +365,10 @@ class SPATIALGDK_API USpatialActorChannel : public UActorChannel
 	// when those properties change.
 	TArray<uint8>* ActorHandoverShadowData;
 	TMap<TWeakObjectPtr<UObject>, TSharedRef<TArray<uint8>>> HandoverShadowDataMap;
+
+	// Band-aid until we get Actor Sets.
+	// Record when an actor got authority from a zoning transition.
+	// If an actor whose position depends on other in the same ownership hierarchy arrives before the others, it could report a position different
+	// from the position it is migrating to. So we prevent the actor from migrating immediately after, we leave some time for the other actors it depends on to arrive.
+	uint64 AuthorityTimestamp;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h:373]: Channels get pooled - this needs to be reset in the `Init` func

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 2]
```diff
@@ -88,6 +88,26 @@ void UpdateChangelistHistory(TUniquePtr<FRepState>& RepState)
...
 	SendingRepState->HistoryEnd = SendingRepState->HistoryStart + NewHistoryCount;
 }
+
+void ForceReplicateOnActorHierarchy(USpatialNetDriver* NetDriver, AActor* RootActor, const AActor* OriginalActor)
+{
+	if (RootActor->GetIsReplicated())
+	{
+		if (RootActor != OriginalActor)
+		{
+			if (USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(NetDriver->PackageMap->GetEntityIdFromObject(RootActor)))
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:98]: Should we just call GetOrCreateSpatialActorChannel here instead?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 3]
```diff
@@ -721,21 +760,44 @@ int64 USpatialActorChannel::ReplicateActor()
...
 	{
 		if (!NetDriver->LoadBalanceStrategy->ShouldHaveAuthority(*Actor) && !NetDriver->LockingPolicy->IsLocked(Actor))
-		{		
-			const VirtualWorkerId NewAuthVirtualWorkerId = NetDriver->LoadBalanceStrategy->WhoShouldHaveAuthority(*Actor);
-			if (NewAuthVirtualWorkerId != SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
+		{
+			AActor* NetOwner = const_cast<AActor*>(Actor->GetNetOwner());
+
+			uint64 HierarchyAuthorityTimestamp = AuthorityTimestamp;
+			if (NetOwner)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:767]:
```suggestion
			if (NetOwner != nullptr)
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 4]
```diff
@@ -721,21 +760,44 @@ int64 USpatialActorChannel::ReplicateActor()
...
 			{
-				Sender->SendAuthorityIntentUpdate(*Actor, NewAuthVirtualWorkerId);
+				GetLatestAuthorityChangeFromHierarchy(NetOwner, HierarchyAuthorityTimestamp);
+			}
 
-				// If we're setting a different authority intent, preemptively changed to ROLE_SimulatedProxy 
-				Actor->Role = ROLE_SimulatedProxy;
-				Actor->RemoteRole = ROLE_Authority;
+			const float TimeSinceReceivingAuthInSeconds = double(FPlatformTime::Cycles64() - HierarchyAuthorityTimestamp) * FPlatformTime::GetSecondsPerCycle64();
+			const float MigrationBackoffTimeInSeconds = 1.0;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:773]:
```suggestion
			const float MigrationBackoffTimeInSeconds = 1.0f;
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 5]
```diff
@@ -721,21 +760,44 @@ int64 USpatialActorChannel::ReplicateActor()
...
-				UE_LOG(LogSpatialActorChannel, Error, TEXT("Load Balancing Strategy returned invalid virtual worker for actor %s"), *Actor->GetName());
+				if (NewAuthVirtualWorkerId != SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
+				{
+					Sender->SendAuthorityIntentUpdate(*Actor, NewAuthVirtualWorkerId);
+
+					// If we're setting a different authority intent, preemptively changed to ROLE_SimulatedProxy 
+					Actor->Role = ROLE_SimulatedProxy;
+					Actor->RemoteRole = ROLE_Authority;
+
+					if (NetOwner)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:790]:
```suggestion
					if (NetOwner != nullptr)
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 6]
```diff
@@ -721,21 +760,44 @@ int64 USpatialActorChannel::ReplicateActor()
...
-				Actor->OnAuthorityLost();
+			const VirtualWorkerId NewAuthVirtualWorkerId = NetDriver->LoadBalanceStrategy->WhoShouldHaveAuthority(*Actor);
+			if (TimeSinceReceivingAuthInSeconds > MigrationBackoffTimeInSeconds)
+			{
+				UE_LOG(LogSpatialActorChannel, Verbose, TEXT("Tried to change auth too early to %i for actor %s"), NewAuthVirtualWorkerId, *Actor->GetName());
 			}
 			else
 			{
-				UE_LOG(LogSpatialActorChannel, Error, TEXT("Load Balancing Strategy returned invalid virtual worker for actor %s"), *Actor->GetName());
+				if (NewAuthVirtualWorkerId != SpatialConstants::INVALID_VIRTUAL_WORKER_ID)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:782]: Should this happen before the migration timer check?
if (workerinvalid)
else if (too soon)
else .. do stuff

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 7]
```diff
@@ -721,21 +760,44 @@ int64 USpatialActorChannel::ReplicateActor()
...
-				Actor->Role = ROLE_SimulatedProxy;
-				Actor->RemoteRole = ROLE_Authority;
+			const float TimeSinceReceivingAuthInSeconds = double(FPlatformTime::Cycles64() - HierarchyAuthorityTimestamp) * FPlatformTime::GetSecondsPerCycle64();
+			const float MigrationBackoffTimeInSeconds = 1.0;
 
-				Actor->OnAuthorityLost();
+			const VirtualWorkerId NewAuthVirtualWorkerId = NetDriver->LoadBalanceStrategy->WhoShouldHaveAuthority(*Actor);
+			if (TimeSinceReceivingAuthInSeconds > MigrationBackoffTimeInSeconds)
+			{
+				UE_LOG(LogSpatialActorChannel, Verbose, TEXT("Tried to change auth too early to %i for actor %s"), NewAuthVirtualWorkerId, *Actor->GetName());
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:778]: Do you see the message spammed much when running back and forth across boundaries?

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h [hunk 8]
```diff
@@ -304,6 +308,8 @@ class SPATIALGDK_API USpatialActorChannel : public UActorChannel
 
 	void InitializeHandoverShadowData(TArray<uint8>& ShadowData, UObject* Object);
 	FHandoverChangeState GetHandoverChangeList(TArray<uint8>& ShadowData, UObject* Object);
+
+	void GetLatestAuthorityChangeFromHierarchy(AActor* RootActor, uint64& OutTimestamp);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h:312]: const?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 9]
```diff
@@ -88,6 +88,26 @@ void UpdateChangelistHistory(TUniquePtr<FRepState>& RepState)
...
+
+void ForceReplicateOnActorHierarchy(USpatialNetDriver* NetDriver, AActor* RootActor, const AActor* OriginalActor)
+{
+	if (RootActor->GetIsReplicated())
+	{
+		if (RootActor != OriginalActor)
+		{
+			if (USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(NetDriver->PackageMap->GetEntityIdFromObject(RootActor)))
+			{
+				Channel->ReplicateActor();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:100]: One problem here is we lose how much data this actor is spending on bandwidth (not something we currently use, however it was teed up to be worked on soon (see channel saturation). This also might not work nicely with the replication graph - would you mind enabling it in the testgyms, and giving your test scenario a spin?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 6 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:782]: I think the timer will actually prevent this error if the actor ends up at an incorrect position because it does not see the actor it is supposed to get its position from.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 7 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:778]: It happens if running in a tight loop. We should have proper hysterisis transitions in the load balancer anyway.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 9 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:100]: On the bandwidth, is this metric still relevant when not acting as a server anymore ?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 2 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:98]: What would it mean to reach that point without having an ActorChannel ? Is it expected to be lazily created ?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 9 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:100]: Tested with TestGym's replication graph enabled. Did not see any issues.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 2 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:98]: Impossible for us to be at the stage of considering authority intent migration without having an Actor channel so there's no need, idm changing if it's a stylistic preference

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 10]
```diff
@@ -88,6 +88,26 @@ void UpdateChangelistHistory(TUniquePtr<FRepState>& RepState)
 	SendingRepState->HistoryStart = SendingRepState->HistoryStart % MaxSendingChangeHistory;
 	SendingRepState->HistoryEnd = SendingRepState->HistoryStart + NewHistoryCount;
 }
+
+void ForceReplicateOnActorHierarchy(USpatialNetDriver* NetDriver, const AActor* RootActor, const AActor* OriginalActor)
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:92]: Shouldn't need to pass in the NetDriver here? SpatialActorChannel already has a reference to it?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 9 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:100]: Any risks with having other hierarchy Actors already in the list of Actors being replicated this tick, this Actor could get replicated twice? any issue with this?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 11]
```diff
@@ -421,6 +442,25 @@ FHandoverChangeState USpatialActorChannel::CreateInitialHandoverChangeState(cons
...
+		if (USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(NetDriver->PackageMap->GetEntityIdFromObject(RootActor)))
+		{
+			if (Channel->AuthorityTimestamp > OutTimestamp)
+			{
+				OutTimestamp = Channel->AuthorityTimestamp;
+			}
+		}
+	}
+
+	for (AActor* Child : RootActor->Children)
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:458]:
```suggestion
	for (const AActor* Child : RootActor->Children)
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 12]
```diff
@@ -721,22 +761,42 @@ int64 USpatialActorChannel::ReplicateActor()
...
+			const AActor* NetOwner = Actor->GetNetOwner();
+
+			uint64 HierarchyAuthorityTimestamp = AuthorityTimestamp;
+			if (NetOwner != nullptr)
+			{
+				GetLatestAuthorityChangeFromHierarchy(NetOwner, HierarchyAuthorityTimestamp);
+			}
+
+			const float TimeSinceReceivingAuthInSeconds = double(FPlatformTime::Cycles64() - HierarchyAuthorityTimestamp) * FPlatformTime::GetSecondsPerCycle64();
+			const float MigrationBackoffTimeInSeconds = 1.0f;
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:774]: Should this be static?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 13]
```diff
@@ -421,6 +442,25 @@ FHandoverChangeState USpatialActorChannel::CreateInitialHandoverChangeState(cons
 	return HandoverChanged;
 }
 
+void USpatialActorChannel::GetLatestAuthorityChangeFromHierarchy(const AActor* RootActor, uint64& OutTimestamp)
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:445]: I don't think `RootActor` is a good name here. It's true for when it's called directly from `ReplicateActor` but then we recursively call it on all Actors in a hierarchy. 

I prefer `HierarchyActor`

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 14]
```diff
@@ -88,6 +88,26 @@ void UpdateChangelistHistory(TUniquePtr<FRepState>& RepState)
 	SendingRepState->HistoryStart = SendingRepState->HistoryStart % MaxSendingChangeHistory;
 	SendingRepState->HistoryEnd = SendingRepState->HistoryStart + NewHistoryCount;
 }
+
+void ForceReplicateOnActorHierarchy(USpatialNetDriver* NetDriver, const AActor* RootActor, const AActor* OriginalActor)
+{
+	if (RootActor->GetIsReplicated())
+	{
+		if (RootActor != OriginalActor)
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:96]: These Ifs can be combined

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 15]
```diff
@@ -88,6 +88,26 @@ void UpdateChangelistHistory(TUniquePtr<FRepState>& RepState)
...
+		{
+			if (USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(NetDriver->PackageMap->GetEntityIdFromObject(RootActor)))
+			{
+				Channel->ReplicateActor();
+			}
+		}
+	}
+
+	for (AActor* Child : RootActor->Children)
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:105]:
```suggestion
	for (const AActor* Child : RootActor->Children)
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 10 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:92]: Also would prefer name changes of `RootActor` -> `HierarchyActor` and `OriginalActor` to `OriginalActorBeingReplicated`

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 16]
```diff
@@ -218,6 +238,7 @@ void USpatialActorChannel::Init(UNetConnection* InConnection, int32 ChannelIndex
 	bIsAuthServer = false;
 	LastPositionSinceUpdate = FVector::ZeroVector;
 	TimeWhenPositionLastUpdated = 0.0f;
+	AuthorityTimestamp = 0;
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:241]: Would pref something more self-explanatory like `TimeSinceAuthorityReceived`

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h [hunk 17]
```diff
@@ -359,4 +365,10 @@ class SPATIALGDK_API USpatialActorChannel : public UActorChannel
 	// when those properties change.
 	TArray<uint8>* ActorHandoverShadowData;
 	TMap<TWeakObjectPtr<UObject>, TSharedRef<TArray<uint8>>> HandoverShadowDataMap;
+
+	// Band-aid until we get Actor Sets.
+	// Record when an actor got authority from a zoning transition.
+	// If an actor whose position depends on other in the same ownership hierarchy arrives before the others, it could report a position different
+	// from the position it is migrating to. So we prevent the actor from migrating immediately after, we leave some time for the other actors it depends on to arrive.
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialActorChannel.h:372]: I found this comment hard to parse, also added a clarification about this being used only on server-side workers for future readers' clarity.

```suggestion
	// Used on server-side workers only.
	// Record when this worker receives SpatialOS Position component authority over the Actor.
	// Tracking this helps prevent authority thrashing which can happen due a replication race 
	// between hierarchy Actors. This happens because hierarchy Actor migration using the 
	// default load-balancing strategy depends on the position of the hierarchy root Actor. 
	// If the hierarchy Actor data is replicated to the new worker before the hierarchy root, it 
	// can immediately attempt to migrate back. Using this timestamp, we can back off attempting
	// migrations for a while.
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 18]
```diff
@@ -721,22 +761,42 @@ int64 USpatialActorChannel::ReplicateActor()
...
 				// If we're setting a different authority intent, preemptively changed to ROLE_SimulatedProxy 
 				Actor->Role = ROLE_SimulatedProxy;
 				Actor->RemoteRole = ROLE_Authority;
 
+				if (NetOwner != nullptr)
+				{
+					ForceReplicateOnActorHierarchy(NetDriver, NetOwner, Actor);
+				}
+
 				Actor->OnAuthorityLost();
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:798]: Have we considered the ordering of setting role to simulated proxy, triggered authority lost, and then doing the same on hierarchy? Might be fine but feel like there could be some issues here

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 12 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:774]: I think this should potentially be exposed in the editor somewhere. When testing this in the example project I was getting confused when my workers weren't changing authority as expected and it was because this backoff was preventing it. That's something I think we can acknowledging affecting here but maybe exposing through editor ( / via lb config?) would provide easier way of testing around it?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 2 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:98]: Wouldn't it just require a new replicated actor that hasn't been replicated yet, before it's inheritance chain attempts to migrate?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 9 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:100]: No but it'll do a lot of unnecessary work. I think if this was reworked so instead of immediately replicating the actor, we signal to the NetDriver or ReplicationGraph that they should include this actor in this frame's replication list, we should solve both the redundant replication, as well as the bandwidth issue.

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:100]: Or use ForceNetUpdate

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 2 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:98]: ah yea, this before I used my brain and realised this function is applied to all hierarchy Actors

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 9 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:100]: Thing is, when migrating we lose Authority on the actor, which will early out of the ReplicateActor (this is also why this code does not result in infinite recursion).

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 12 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:774]: I agree, I'll put that in a follow up task.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 2 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:98]: Cool - so I think we should swap this with `GetOrCreateSpatialActorChannel`

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 9 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:100]: Discussed offline - will create a follow up ticket to track this work

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 18 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Yes, it would be more consistent.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 12 above)
Comment (jennifer-at-improbable-io) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:771]: Do we want to be storing the Authority change in the SpatialActorChannel or in the LBStrategy? I'm happy to have that in a followup as well.

Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:771]: https://improbableio.atlassian.net/browse/UNR-3335 for time setting

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 9 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: https://improbableio.atlassian.net/browse/UNR-3336

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp (hunk 12 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:771]: > Do we want to be storing the Authority change in the SpatialActorChannel or in the LBStrategy? I'm happy to have that in a followup as well.

Despite there not being use for this authority timestamp outside load-balancing, I think it's fine to store it here given how authority role is an Actor-level concept in Unreal.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 19]
```diff
@@ -721,21 +758,45 @@ int64 USpatialActorChannel::ReplicateActor()
...
+					Actor->Role = ROLE_SimulatedProxy;
+					Actor->RemoteRole = ROLE_Authority;
+
+					Actor->OnAuthorityLost();
+
+					if (NetOwner != nullptr)
+					{
+						ForceReplicateOnActorHierarchy(NetDriver, NetOwner, Actor);
+					}
+
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:798]:
```suggestion
```
